x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushFiltersToSource.java

```diff
@@ -53,45 +59,94 @@
-                queryExec = new EsQueryExec(
-                    queryExec.source(),
-                    queryExec.index(),
-                    queryExec.indexMode(),
-                    queryExec.output(),
-                    query,
-                    queryExec.limit(),
-                    queryExec.sorts(),
-                    queryExec.estimatedRowSize()
-                );
-                if (nonPushable.size() > 0) { // update filter with remaining non-pushable conditions
-                    plan = new FilterExec(filterExec.source(), queryExec, Predicates.combineAnd(nonPushable));
-                } else { // prune Filter entirely
-                    plan = queryExec;
-                }
-            } // else: nothing changes
+        });
+        AttributeMap<Attribute> aliasReplacedBy = aliasReplacedByBuilder.build();
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            (canPushToSource(resExp, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats())) ? pushable : nonPushable).add(
+                exp
+            );
         }
+        // Replace field references with their actual field attributes
+        pushable.replaceAll(e -> e.transformDown(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r)));
+        return rewrite(filterExec, queryExec, pushable, nonPushable, others);
+    }
 
-        return plan;
+    private static PhysicalPlan rewrite(
+        FilterExec filterExec,
+        EsQueryExec queryExec,
+        List<Expression> pushable,
+        List<Expression> nonPushable,
+        List<Alias> others
+    ) {
+        // Combine GT, GTE, LT and LTE in pushable to Range if possible
+        List<Expression> newPushable = combineEligiblePushableToRange(pushable);
+        if (newPushable.size() > 0) { // update the executable with pushable conditions
+            Query queryDSL = PlannerUtils.TRANSLATOR_HANDLER.asQuery(Predicates.combineAnd(newPushable));
+            QueryBuilder planQuery = queryDSL.asBuilder();
+            var query = Queries.combine(Queries.Clause.FILTER, asList(queryExec.query(), planQuery));
+            queryExec = new EsQueryExec(
+                queryExec.source(),
+                queryExec.index(),
+                queryExec.indexMode(),
+                queryExec.output(),
+                query,
+                queryExec.limit(),
+                queryExec.sorts(),
+                queryExec.estimatedRowSize()
+            );
+            // If the eval contains other aliases, not just field attributes, we need to keep them in the plan
+            PhysicalPlan plan = others.isEmpty() ? queryExec : new EvalExec(filterExec.source(), queryExec, others);
```

Comment:
Indeed. Once I found a failing test, I fixed this, and then found some existing tests that failed because they did not actually assert on the right thing, so this was a bug affecting more cases than I realized. Our integration tests did not notice because they usually have a KEEP, and that always modifies the plan at logical planning time to avoid this bug. But queries with EVAL and no KEEP quite easily hit this.